### ISSUE TITLE
Variable interpolation fallback broken

### ISSUE THREAD

**Issue by Asherlc** [serverless/serverless#5205]
# This is a Bug Report

## Description

With a nested variable interpolation with fallbacks, they seem to evaluate to false no matter what. I.e.:  `${self:custom.foo.${opt:stage, self:provider.stage}, 'shouldNotSee'}` will *always* result in `shouldNotSee`. This is a regression introduced in 1.30.

A replication is here: https://github.com/Asherlc/serverless-interpolation-bug

Run `yarn serverless print --stage production` to see the issue. Reverting to 1.29 solves the issue.

* ***Serverless Framework Version you're using***: 1.30
* ***Operating System***: macOS 10.13.6
* ***Stack Trace***: N/A
* ***Provider Error messages***: N/A

**Comment by rob-gonz** [comment]
Can confirm, I am seeing this exact behavior.  I just upgraded from 1.29 to 1.30 and all my variables based on this interpolation have ceased to function properly.  I'll have to revert to 1.29.

**Comment by dschep** [comment]
I think it also introduced this issue I'm seeing:

`serverless.yml`:
```yaml
service:
  name: iam-error-test
  publish: false
custom:
  region: us-east-1
  stage: ${opt:stage, self:provider.stage}
provider:
  name: aws
  # Prevents serverless from trying to sub out ${AWS::blah}
  variableSyntax: "\\${((?!AWS)[ ~:a-zA-Z0-9._'\",\\-\\/\\(\\)]+?)}"
  region: ${self:custom.region}
  versionFunctions: false
  stage: ${env:USER, 'dev'}
  runtime: python3.6
  iamRoleStatements:
    - Effect: Allow
      Action:
        - lambda:InvokeFunction
      # This is effectively saying:
      # Resource: arn:aws:lambda:${self:custom.region}:${AWS::AccountId}:function:*
      Resource:
        - {'Fn::Join': [':', ['arn:aws:lambda', {Ref: 'AWS::Region'}, {Ref: 'AWS::AccountId'}, 'function:trading-${self:custom.${self:custom.stage}.stage, self:custom.stage}-*']]}

functions:
  test:
    handler: handlers.test
    events:
      - http:
          path: test
          method: get

```

`sls print` with 1.29.2 & older:
```yaml
service:
  name: iam-error-test
  publish: false
custom:
  region: us-east-1
  stage: dschep
provider:
  stage: dschep
  region: us-east-1
  name: aws
  versionFunctions: false
  runtime: python3.6
  iamRoleStatements:
    - Effect: Allow
      Action:
        - 'lambda:InvokeFunction'
      Resource:
        - 'Fn::Join':
            - ':'
            - - 'arn:aws:lambda'
              - Ref: 'AWS::Region'
              - Ref: 'AWS::AccountId'
              - 'function:trading-dschep-*'
  variableSyntax: '\${((?!AWS)[ ~:a-zA-Z0-9._''",\-\/\(\)]+?)}'
functions:
  test:
    handler: handlers.test
    events:
      - http:
          path: test
          method: get

```
`sls print` with 1.30.0:
```yaml
service:
  name: iam-error-test
  publish: false
custom:
  region: us-east-1
  stage: dschep
provider:
  stage: dschep
  region: us-east-1
  name: aws
  versionFunctions: false
  runtime: python3.6
  iamRoleStatements:
    - Effect: Allow
      Action:
        - 'lambda:InvokeFunction'
      Resource:
        - 'Fn::Join':
            - ':'
            - - 'arn:aws:lambda'
              - Ref: 'AWS::Region'
              - Ref: 'AWS::AccountId'
              - >-
                function:trading-function:trading-function:trading-function:trading-dschep-*-*-*-*
  variableSyntax: '\${((?!AWS)[ ~:a-zA-Z0-9._''",\-\/\(\)]+?)}'
functions:
  test:
    handler: handlers.test
    events:
      - http:
          path: test
          method: get

```


where the `-*-*-*-*` is obviously wrong and should just be `-*`

**Comment by bugbreaka** [comment]
Just updated to 1.30.0; deploy worked fine without --stage argument, but things got wonky after giving `--stage qa`.

Here is a snippet from my serverless.yaml:

```
service: service-name

provider:
  stage: ${opt:stage, 'dev'}
  environment:
    TABLE_NAME: ${self:service}-${opt:stage, self:provider.stage}-table

resources:
  Description: CloudFormation stack for ${self:service}-${opt:stage, self:provider.stage}.
```

The TABLE_NAME with the `--stage qa` argument became `service-name-service-name-qa-table-table`, also the Description had repeated content.

The deploy worked fine after reverting to 1.29.2 version.

**Comment by dschep** [comment]
That's odd @bugbreaka, it's the same duplication issue I'm seeing, but I'm not using `-s`.

**Comment by erikerikson** [comment]
Looking into this.  Apologies for the issue.

**Comment by erikerikson** [comment]
This occurs because although it does in most cases, in others `match.match != property` at https://github.com/serverless/serverless/blob/master/lib/classes/Variables.js#L304 . Coding a test and a fix.

**Comment by erikerikson** [comment]
Fix offered on #5224.  Please try it out and be sure to let me know if I missed anything in your config.  We continue building out the case coverage.  Thanks for reporting and apologies for the friction!
